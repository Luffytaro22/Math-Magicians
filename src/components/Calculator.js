/* eslint-disable react/prop-types */
import React, { useState } from 'react';
import calculate from '../logic/calculate';

// Child component
function CalculatorButton({
  className, text, id, obj, setObj,
}) {
  const handleClick = (event) => {
    const operation = calculate(obj, event.target.textContent);

    setObj((oldObj) => ({
      ...oldObj,
      ...operation,
    }));

    const result = document.getElementById('result');
    // If there's a zero, update the content.
    if (result.textContent === '0') {
      result.textContent = '';
    }
    // Concat the operations.
    result.textContent += event.target.textContent;
    // If the user press equal, display the result.
    if (event.target.textContent === '=') {
      result.textContent = '';
      result.textContent = operation.total || '0';
    } else if (event.target.textContent === 'AC') {
      result.textContent = '0';
    }
  };

  return (
    <button type="button" className={className} id={id} onClick={handleClick}>
      {text}
    </button>
  );
}

// Main calculator component
export default function Calculator() {
  const [obj, setObj] = useState({ total: null, next: null, operation: null });

  return (
    <div id="calculator-container">
      <div id="result">0</div>
      <div id="operations-container">
        <CalculatorButton className="operations" text="AC" obj={obj} setObj={setObj} />
        <CalculatorButton className="operations" text="+/-" obj={obj} setObj={setObj} />
        <CalculatorButton className="operations" text="%" obj={obj} setObj={setObj} />
        <CalculatorButton className="operations right-operations" text="รท" obj={obj} setObj={setObj} />
        <CalculatorButton className="operations" text="7" obj={obj} setObj={setObj} />
        <CalculatorButton className="operations" text="8" obj={obj} setObj={setObj} />
        <CalculatorButton className="operations" text="9" obj={obj} setObj={setObj} />
        <CalculatorButton className="operations right-operations" text="x" obj={obj} setObj={setObj} />
        <CalculatorButton className="operations" text="4" obj={obj} setObj={setObj} />
        <CalculatorButton className="operations" text="5" obj={obj} setObj={setObj} />
        <CalculatorButton className="operations" text="6" obj={obj} setObj={setObj} />
        <CalculatorButton className="operations right-operations" text="-" obj={obj} setObj={setObj} />
        <CalculatorButton className="operations" text="1" obj={obj} setObj={setObj} />
        <CalculatorButton className="operations" text="2" obj={obj} setObj={setObj} />
        <CalculatorButton className="operations" text="3" obj={obj} setObj={setObj} />
        <CalculatorButton className="operations right-operations" text="+" obj={obj} setObj={setObj} />
        <CalculatorButton className="operations" id="zero-button" text="0" obj={obj} setObj={setObj} />
        <CalculatorButton className="operations" text="." obj={obj} setObj={setObj} />
        <CalculatorButton className="operations right-operations" text="=" obj={obj} setObj={setObj} />
      </div>
    </div>
  );
}
