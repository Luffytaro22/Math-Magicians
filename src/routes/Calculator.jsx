import Calculator from '../components/Calculator';

export default function CalculatorPage() {
  return (
    <div id="calculator-div">
      <h2>Let&aposs do some math!</h2>
      <Calculator />
    </div>
  );
}
