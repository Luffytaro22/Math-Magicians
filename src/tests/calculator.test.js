import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Calculator from '../components/Calculator';

test('renders Calculator correctly', () => {
  const { container } = render(<Calculator />);
  expect(container).toMatchSnapshot();
});

test('updates result on button click', () => {
  const { getByText } = render(<Calculator />);
  const result = document.getElementById('result');

  fireEvent.click(getByText('1'));
  fireEvent.click(getByText('+'));
  fireEvent.click(getByText('2'));
  fireEvent.click(getByText('='));

  expect(result.textContent).toBe('3');

  // Update the snapshot
  expect(result).toMatchSnapshot();
});

test('clears result on "AC" button click', () => {
  const { getByText } = render(<Calculator />);
  const result = document.getElementById('result');

  fireEvent.click(getByText('1'));
  fireEvent.click(getByText('+'));
  fireEvent.click(getByText('2'));
  fireEvent.click(getByText('AC'));

  expect(result.textContent).toBe('0');

  // Update the snapshot
  expect(result).toMatchSnapshot();
});

test('handles division by zero', () => {
  const { getByText } = render(<Calculator />);
  const result = document.getElementById('result');

  fireEvent.click(getByText('5'));
  fireEvent.click(getByText('รท'));
  fireEvent.click(getByText('0'));
  fireEvent.click(getByText('='));

  expect(result.textContent).toBe("Can't divide by 0.");

  // Update the snapshot
  expect(result).toMatchSnapshot();
});
