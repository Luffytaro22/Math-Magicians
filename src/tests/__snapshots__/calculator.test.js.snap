// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clears result on "AC" button click 1`] = `
<div
  id="result"
>
  0
</div>
`;

exports[`handles division by zero 1`] = `
<div
  id="result"
>
  Can't divide by 0.
</div>
`;

exports[`renders Calculator correctly 1`] = `
<div>
  <div
    id="calculator-container"
  >
    <div
      id="result"
    >
      0
    </div>
    <div
      id="operations-container"
    >
      <button
        class="operations"
        id="AC"
        type="button"
      >
        AC
      </button>
      <button
        class="operations"
        id="+/-"
        type="button"
      >
        +/-
      </button>
      <button
        class="operations"
        id="%"
        type="button"
      >
        %
      </button>
      <button
        class="operations right-operations"
        id="รท"
        type="button"
      >
        รท
      </button>
      <button
        class="operations"
        id="7"
        type="button"
      >
        7
      </button>
      <button
        class="operations"
        id="8"
        type="button"
      >
        8
      </button>
      <button
        class="operations"
        id="9"
        type="button"
      >
        9
      </button>
      <button
        class="operations right-operations"
        id="x"
        type="button"
      >
        x
      </button>
      <button
        class="operations"
        id="4"
        type="button"
      >
        4
      </button>
      <button
        class="operations"
        id="5"
        type="button"
      >
        5
      </button>
      <button
        class="operations"
        id="6"
        type="button"
      >
        6
      </button>
      <button
        class="operations right-operations"
        id="-"
        type="button"
      >
        -
      </button>
      <button
        class="operations"
        id="1"
        type="button"
      >
        1
      </button>
      <button
        class="operations"
        id="2"
        type="button"
      >
        2
      </button>
      <button
        class="operations"
        id="3"
        type="button"
      >
        3
      </button>
      <button
        class="operations right-operations"
        id="+"
        type="button"
      >
        +
      </button>
      <button
        class="operations"
        id="zero-button"
        type="button"
      >
        0
      </button>
      <button
        class="operations"
        id="."
        type="button"
      >
        .
      </button>
      <button
        class="operations right-operations"
        id="="
        type="button"
      >
        =
      </button>
    </div>
  </div>
</div>
`;

exports[`updates result on button click 1`] = `
<div
  id="result"
>
  3
</div>
`;
